sql/colexec: TestDistinct failed

[(sql/colexec).TestDistinct failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1999346&tab=buildLog) on [master@3c756942000f7c2292ed1c3e68f78aa1d9f33dd2](https://github.com/cockroachdb/cockroach/commits/3c756942000f7c2292ed1c3e68f78aa1d9f33dd2):

```
=== RUN   TestDistinct
--- FAIL: TestDistinct (0.11s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=false (0.00s)
            utils_test.go:382: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=true (0.00s)
            utils_test.go:382: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01
    --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01 (0.00s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=false (0.00s)
            utils_test.go:382: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=true (0.00s)
            utils_test.go:382: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestDistinct PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDistinct.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colexec).TestDistinct failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2012270&tab=buildLog) on [master@13356765ce742e90a1abd274f7862d46f687cead](https://github.com/cockroachdb/cockroach/commits/13356765ce742e90a1abd274f7862d46f687cead):

```
=== RUN   TestDistinct
--- FAIL: TestDistinct (0.22s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1/batchSize=3/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1/batchSize=3/sel=false (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1/batchSize=3/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1/batchSize=3/sel=true (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1
    --- FAIL: TestDistinct/partiallyOrdered/ordCols=1 (0.00s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=true (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=false (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01
    --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01 (0.00s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1/batchSize=1/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1/batchSize=1/sel=false (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1/batchSize=1/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1/batchSize=1/sel=true (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=false (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=true (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestDistinct PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDistinct.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colexec).TestDistinct failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2014105&tab=buildLog) on [master@9eea0be192499fa7d29b5eca5ed14173982b25f0](https://github.com/cockroachdb/cockroach/commits/9eea0be192499fa7d29b5eca5ed14173982b25f0):

```
=== RUN   TestDistinct
--- FAIL: TestDistinct (0.11s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=true (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01
    --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01 (0.00s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=false (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=true (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=false (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestDistinct PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDistinct.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colexec).TestDistinct failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2014106&tab=buildLog) on [master@9eea0be192499fa7d29b5eca5ed14173982b25f0](https://github.com/cockroachdb/cockroach/commits/9eea0be192499fa7d29b5eca5ed14173982b25f0):

```
=== RUN   TestDistinct
--- FAIL: TestDistinct (0.33s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01
    --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01 (0.01s)
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=false (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=1/sel=true (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=false
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=false (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]
=== RUN   TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=true
        --- FAIL: TestDistinct/partiallyOrdered/ordCols=1#01/batchSize=3/sel=true (0.00s)
            utils_test.go:383: expected [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4]], actual [[<nil>, <nil>, <nil>, <nil>], [<nil>, <nil>, 30, <nil>], [1, 2, 30, 4], [2, 2, 30, 4], [2, 3, 30, 4], [2, 3, 40, 4], [2, 3, 40, 4]]

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestDistinct PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDistinct.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/hashtable.go
+++ b/pkg/sql/colexec/hashtable.go
@@ -480,14 +480,22 @@ func (ht *hashTable) reset(_ context.Context) {
 	}
 	ht.vals.ResetInternalBatch()
 	ht.vals.SetLength(0)
-	// ht.next, ht.same and ht.visited are reset separately before
+	// ht.probeScratch.next, ht.same and ht.visited are reset separately before
 	// they are used (these slices are not used in all of the code paths).
-	// ht.buckets doesn't need to be reset because buckets are always initialized
-	// when computing the hash.
+	// ht.probeScratch.buckets doesn't need to be reset because buckets are
+	// always initialized when computing the hash.
 	copy(ht.probeScratch.groupID[:coldata.BatchSize()], zeroUint64Column)
-	// ht.toCheck doesn't need to be reset because it is populated manually every
-	// time before checking the columns.
+	// ht.probeScratch.toCheck doesn't need to be reset because it is populated
+	// manually every time before checking the columns.
 	copy(ht.probeScratch.headID[:coldata.BatchSize()], zeroUint64Column)
 	copy(ht.probeScratch.differs[:coldata.BatchSize()], zeroBoolColumn)
 	copy(ht.probeScratch.distinct, zeroBoolColumn)
+	if ht.buildMode == hashTableDistinctBuildMode && cap(ht.buildScratch.next) > 0 {
+		// In "distinct" build mode, ht.buildScratch.next is populated
+		// iteratively, whenever we find tuples that we haven't seen before. In
+		// order to reuse the same underlying memory we need to slice up that
+		// slice (note that keyID=0 is reserved for the end of all hash chains,
+		// so we make the length 1).
+		ht.buildScratch.next = ht.buildScratch.next[:1]
+	}
 }

